### PR TITLE
Added NVarchar2 DataType for oracle utf8 text data handling (when ins…

### PR DESCRIPTION
**Pull request by lvaleriu**
…erting characters that don't belong to the national char set)

Example:
```cs
       using (var db = new DataConnection("db"))
            {
                IQueryable<Data> query= from p in db.GetTable<Data>()
                    where p.Id == 1
                    select p;

                Data res = query.FirstOrDefault();

                string title = "西格玛西格玛西格玛西格玛";

                Assert.That(res.Title, Is.Not.EqualTo(title));

                query.Set(e => e.Title, () => title)
                    .Update();

                res = (from p in db.GetTable<Data>()
                    where p.Id == res.Id
                    select p).FirstOrDefault();

                Assert.That(res.Title, Is.EqualTo(title));
            }
```
```cs
[Table(Name = "DATA", Schema = "schema")]
public class Data:
{
        [Column("Id"), PrimaryKey, SequenceName("schema" + ".SEQ_DATA"), Identity]
        public int Id{ get; set; }

        [Column("TITLE", DataType = DataType.NVarchar2)]
        public string Title { get; set; }
}
```

### CODE REVIEW

**Comment by sdanyliv** [comment]
Please correct leading spaces. They should be replaced by tabs.

**Comment by lvaleriu** [comment]
Is it ok now?

**Comment by sdanyliv** [comment]
Great, will check code and test tomorrow. Thank you.

**Comment by lvaleriu** [comment]
Oracle database settings:
image

Title column (it is called URL in our database in reality, but anyway):

image

**Comment by sdanyliv** [comment]
@lvaleriu, can you provide unit test? Looks like special case needed. I'm not so familiar with Oracle.

**Comment by lvaleriu** [comment]
I can create the unit tests but have no idea how to modify the oracle.sql script. The idea is to set these parameters for the oracle database.

setenv NLS_CHARACTERSET WE8ISO8859P15
setenv NLS_NCHAR_CHARACTERSET AL16UTF16

**Comment by sdanyliv** [comment]
Please create unit tests, i'll try to setup my virtual mashine to handle these settings.

File: Source/DataType.cs [hunk 1]
```diff
@@ -33,6 +33,27 @@ public enum DataType
...
+		/// national character set. AL16UTF16 and UTF8 are both Unicode encoding. The NCHAR datatype stores fixed-length 
+		/// character strings that correspond to the national character set.The NVARCHAR2 datatype stores variable length 
+		/// character strings. When you create a table with an NCHAR or NVARCHAR2 column, the maximum size specified is 
+		/// always in character length semantics. Character length semantics is the default and only length semantics for
+		///  NCHAR or NVARCHAR2. For example, if national character set is UTF8, then the following statement defines the
+		///  maximum byte length of 90 bytes: CREATE TABLE tab1 (col1 NCHAR(30)); This statement creates a column with 
+		/// maximum character length of 30. The maximum byte length is the multiple of the maximum character length and 
+		/// the maximum number of bytes in each character.
+		/// The maximum length of an NVARCHAR2 column is 4000 bytes. It can hold up to 4000 characters. The actual data
+		///  is subject to the maximum byte limit of 4000. The two size constraints must be satisfied simultaneously at run time.
```
Comment (sdanyliv) [inline Source/DataType.cs]: Maybe more shorter description with link to documentation?

**Comment by sdanyliv** [comment]
Will return to this PR on next week.

**Comment by ili** [comment]
Test fails on my machine:
```
Oracle.ManagedDataAccess.Client.OracleException : Unsupported column datatype
   в Oracle.ManagedDataAccess.Client.OracleParameter.PreBind(OracleConnectionImpl connImpl, ColumnDescribeInfo cachedParamMetadata, Boolean& bMetadataModified, Int32 arrayBindCount, ColumnDescribeInfo& paramMetaData, Object& paramValue, Boolean isEFSelectStatement, SqlStatementType stmtType)
   в OracleInternal.ServiceObjects.OracleCommandImpl.InitializeParamInfo(ICollection paramColl, OracleConnectionImpl connectionImpl, ColumnDescribeInfo[] cachedParamMetadata, Boolean& bMetadataModified, Boolean isEFSelectStatement, MarshalBindParameterValueHelper& marshalBindValuesHelper)
   в OracleInternal.ServiceObjects.OracleCommandImpl.ProcessParameters(OracleParameterCollection paramColl, OracleConnectionImpl connectionImpl, ColumnDescribeInfo[] cachedParamMetadata, Boolean& bBindMetadataModified, Boolean isEFSelectStatement, MarshalBindParameterValueHelper& marshalBindValuesHelper)
   в OracleInternal.ServiceObjects.OracleCommandImpl.ExecuteNonQuery(String commandText, OracleParameterCollection paramColl, CommandType commandType, OracleConnectionImpl connectionImpl, Int32 longFetchSize, Int64 clientInitialLOBFS, OracleDependencyImpl orclDependencyImpl, Int64[]& scnFromExecution, OracleParameterCollection& bindByPositionParamColl, Boolean& bBindParamPresent, OracleException& exceptionForArrayBindDML, Boolean isFromEF)
   в Oracle.ManagedDataAccess.Client.OracleCommand.ExecuteNonQuery()
   в LinqToDB.Data.DataConnection.ExecuteNonQuery() в F:\job\linq2db\Source\Data\DataConnection.cs:строка 667
   в LinqToDB.Data.DataConnection.LinqToDB.IDataContext.ExecuteNonQuery(Object query) в F:\job\linq2db\Source\Data\DataConnection.Linq.cs:строка 175
   в LinqToDB.Linq.Query`1.NonQueryQuery(IDataContextInfo dataContextInfo, Expression expr, Object[] parameters) в F:\job\linq2db\Source\Linq\Query.cs:строка 253
   в LinqToDB.Linq.Query`1.<SetNonQueryQuery>b__16_0(QueryContext ctx, IDataContextInfo db, Expression expr, Object[] ps) в F:\job\linq2db\Source\Linq\Query.cs:строка 240
   в LinqToDB.Linq.Query`1.Insert(IDataContextInfo dataContextInfo, T obj, String tableName, String databaseName, String schemaName) в F:\job\linq2db\Source\Linq\Query.cs:строка 634
   в LinqToDB.DataExtensions.Insert[T](IDataContext dataContext, T obj, String tableName, String databaseName, String schemaName) в F:\job\linq2db\Source\DataExtensions.cs:строка 197
   в Tests.DataProvider.OracleTests.NVarchar2InsertTest(String context) в F:\job\linq2db\Tests\Linq\DataProvider\OracleTests.cs:строка 661
```
```sql
DataConnection: -- Oracle.Managed Oracle
DELETE FROM
	TESTUSER.ALLTYPES t1
WHERE
	t1.ID >= 1000

DataConnection: Execution time: 00:00:00.0010001. Records affected: 0.

DataConnection: -- Oracle.Managed Oracle
DECLARE @ID Decimal(4,0)
SET     @ID = 1000
DECLARE @BIGINTDATATYPE Decimal
SET     @BIGINTDATATYPE = NULL
DECLARE @NUMERICDATATYPE Decimal
SET     @NUMERICDATATYPE = NULL
DECLARE @BITDATATYPE Decimal
SET     @BITDATATYPE = NULL
DECLARE @SMALLINTDATATYPE Decimal
SET     @SMALLINTDATATYPE = NULL
DECLARE @DECIMALDATATYPE Decimal
SET     @DECIMALDATATYPE = NULL
DECLARE @SMALLMONEYDATATYPE Decimal
SET     @SMALLMONEYDATATYPE = NULL
DECLARE @INTDATATYPE Decimal
SET     @INTDATATYPE = NULL
DECLARE @TINYINTDATATYPE Decimal
SET     @TINYINTDATATYPE = NULL
DECLARE @MONEYDATATYPE Decimal
SET     @MONEYDATATYPE = NULL
DECLARE @FLOATDATATYPE BinaryDouble -- Double
SET     @FLOATDATATYPE = NULL
DECLARE @REALDATATYPE BinaryFloat -- Single
SET     @REALDATATYPE = NULL
DECLARE @DATETIMEDATATYPE Date
SET     @DATETIMEDATATYPE = NULL
DECLARE @DATETIME2DATATYPE TimeStamp -- DateTime
SET     @DATETIME2DATATYPE = NULL
DECLARE @DATETIMEOFFSETDATATYPE TimeStampTZ -- DateTime
SET     @DATETIMEOFFSETDATATYPE = NULL
DECLARE @LOCALZONEDATATYPE TimeStampTZ -- DateTime
SET     @LOCALZONEDATATYPE = NULL
DECLARE @CHARDATATYPE Char -- AnsiStringFixedLength
SET     @CHARDATATYPE = NULL
DECLARE @VARCHARDATATYPE Varchar2 -- String
SET     @VARCHARDATATYPE = NULL
DECLARE @TEXTDATATYPE Clob -- Object
SET     @TEXTDATATYPE = NULL
DECLARE @NCHARDATATYPE Char -- StringFixedLength
SET     @NCHARDATATYPE = NULL
DECLARE @NVARCHARDATATYPE NVarchar2(8) -- String
SET     @NVARCHARDATATYPE = '致我们最爱的母亲'
DECLARE @NTEXTDATATYPE NClob -- Object
SET     @NTEXTDATATYPE = NULL
DECLARE @BINARYDATATYPE Raw -- Binary
SET     @BINARYDATATYPE = NULL
DECLARE @BFILEDATATYPE Blob -- Object
SET     @BFILEDATATYPE = NULL
DECLARE @GUIDDATATYPE Blob -- Object
SET     @GUIDDATATYPE = NULL
DECLARE @URIDATATYPE 0 -- Object
SET     @URIDATATYPE = NULL
DECLARE @XMLDATATYPE Varchar2 -- String
SET     @XMLDATATYPE = NULL

INSERT INTO TESTUSER.ALLTYPES
(
	ID,
	BIGINTDATATYPE,
	NUMERICDATATYPE,
	BITDATATYPE,
	SMALLINTDATATYPE,
	DECIMALDATATYPE,
	SMALLMONEYDATATYPE,
	INTDATATYPE,
	TINYINTDATATYPE,
	MONEYDATATYPE,
	FLOATDATATYPE,
	REALDATATYPE,
	DATETIMEDATATYPE,
	DATETIME2DATATYPE,
	DATETIMEOFFSETDATATYPE,
	LOCALZONEDATATYPE,
	CHARDATATYPE,
	VARCHARDATATYPE,
	TEXTDATATYPE,
	NCHARDATATYPE,
	NVARCHARDATATYPE,
	NTEXTDATATYPE,
	BINARYDATATYPE,
	BFILEDATATYPE,
	GUIDDATATYPE,
	URIDATATYPE,
	XMLDATATYPE
)
VALUES
(
	:ID,
	:BIGINTDATATYPE,
	:NUMERICDATATYPE,
	:BITDATATYPE,
	:SMALLINTDATATYPE,
	:DECIMALDATATYPE,
	:SMALLMONEYDATATYPE,
	:INTDATATYPE,
	:TINYINTDATATYPE,
	:MONEYDATATYPE,
	:FLOATDATATYPE,
	:REALDATATYPE,
	:DATETIMEDATATYPE,
	:DATETIME2DATATYPE,
	:DATETIMEOFFSETDATATYPE,
	:LOCALZONEDATATYPE,
	:CHARDATATYPE,
	:VARCHARDATATYPE,
	:TEXTDATATYPE,
	:NCHARDATATYPE,
	:NVARCHARDATATYPE,
	:NTEXTDATATYPE,
	:BINARYDATATYPE,
	:BFILEDATATYPE,
	:GUIDDATATYPE,
	:URIDATATYPE,
	:XMLDATATYPE
)
```

File: Source/DataType.cs (hunk 1 above)
Comment (lvaleriu) [inline Source/DataType.cs]: Should i put the documentation somewhere on the repository?

**Comment by lvaleriu** [comment]
I'm using Oracle.DataAccess for this feature. Let me try to use Oracle.ManagedDataAccess client.

**Comment by lvaleriu** [comment]
I've tested with Oracle.ManagedDataAccess dll and it works. What kind of Oracle Server installation you have on your virtual machine?

Have you installed such things? http://www.oracle.com/technetwork/database/windows/downloads/index-090165.html

**Comment by sdanyliv** [comment]
Will check on next week, thanks for update.

**Comment by sdanyliv** [comment]
Tests in our environment are failed.
```
Personal Oracle Database 11g Release 11.2.0.1.0 - 64bit Production
PL/SQL Release 11.2.0.1.0 - Production
"CORE	11.2.0.1.0	Production"
TNS for 64-bit Windows: Version 11.2.0.1.0 - Production
NLSRTL Version 11.2.0.1.0 - Production
```

**Comment by lvaleriu** [comment]
Thanks!

**Pull request merged**